Revamp volumes documentation (#1221)
docksal/docksal: docs/content/core/volumes.md

| macOS   | Docker Desktop  | [bind](#bind)                | **Pros:** supports filesystem events. <br> **Cons:** pretty slow, 40% slower than native filesystem. |
| Windows | ANY             | [bind](#bind) (**default**)  | **Pros:** ~20% overhead as compared to native FS. <br> **Cons:** does not support filesystem events (fsnotify). |
| macOS, Windows | ANY      | [unison](#unison)            | **Pros:** maximum `cli` filesystem performance. <br> **Cons:** initial wait for files to sync into `cli`; additional Docksal disk space use; sync delay when you switch git branches; higher CPU usage during files sync; sometimes Unison might 'break.' |
| ANY     | ANY             | [none](#none)                | **Pros:** maximum `cli` filesystem performance and no wait for the initial sync. <br> **Cons:** you have to copy files manually or checkout and edit files inside `cli` container. |

## Project Volumes

If you are familiar with Docker Compose, then you must have attached files or dirs like this:

```yaml
  cli:
    volumes:
      - /Users/alex/mysite:/var/www:rw
```

> It is an unnamed volume with bind driver. However to support various OS in Docksal we cannot just always rely on the 

Author: In Russian it sounds alright xD I think I meant "we just cannot always"

> bind driver, and should not use unnamed volumes. 

Author: What about "always comma before and" rule?

- `db_data` to store your database data independently of the `db` container 
- `docksal_ssh_agent` to share SSH keys with containers

Docksal made these volumes to function transparently across different operating systems, but their actual definitions 
vary depending on the `DOCKSAL_VOLUMES` option.

{{% notice note %}}
`project_root` volume is mapped to a specific location on the host. 
If you need to move the location of a project on the computer, remove the project stack first with
`fin project rm`, then you can move the project folder and start the project stack again.
{{% /notice %}}

{{% notice warning %}}
When customizing `docksal.yml` make sure to include the `cached` option **anywhere** the `project_root` volume is attached
to a service. Mixing `cached` and non-`cached` mounts for the same volume in your project stack will lead to 
issues and errors with Docker Desktop on macOS. See [docksal/docksal#678](https://github.com/docksal/docksal/issues/678) for more details.
{{% /notice %}}

## DOCKSAL_VOLUMES

`DOCKSAL_VOLUMES` value changes the driver option for project volumes mentioned above, and also affects some additional `fin` behavior. 

### bind 

> With this option containers access files using **Docker bind** driver option, which basically means direct access. 

Author: sorry?

> With this option containers access files using **Docker bind** driver option, which basically means direct access. 

Reviewer: Comma after option: With this option, containers

> With this option containers access files using **Docker bind** driver option, which basically means direct access. 

Author: There is another word option that has comma after it already, so comment was confising

> While Docker can access files directly on Linux, on macOS and Windows it works inside VM (VirtualBox or xhyve with Docker Desktop), 

Author: it's not a specific VM in this case

> While Docker can access files directly on Linux, on macOS and Windows it works inside VM (VirtualBox or xhyve with Docker Desktop), 

Reviewer: It is specific as far as it is the VM operating. “A” VM (for something generic) doesn’t work because you are specifying an operation. It’s just a quirky American English thing that requires a definite article.

```

**On Windows with Docker Desktop** it is Docker Desktop itself that mounts all configured Windows drives via SMB.

```
               SMB mount                       bind
Windows Host =============> Docker Desktop ==============> Container   
```

To see how your project's Docker volumes are defined with `DOCKSAL_VOLUMES=bind` see 
[stacks/volumes-bind.yml](https://github.com/docksal/docksal/blob/master/stacks/volumes-bind.yml).

In most cases, you do not need to set `DOCKSAL_VOLUES=bind` option. It is set for you automatically 
in applicable cases. The only thinkable exception is when you need fsnotify events on macOS with Docker Desktop, 
but don't want to use `unison` option.

### nfs

This option is macOS specific and is used by default on macOS instead of bind. It means that containers will not wait 
for something to make host files accessible for them first – instead they will reach out to host themselves via NFS protocol.

```
              NFS mount
macOS Host ==============> Container  
```

NFS generally works faster than `osxfs`, so this option is default on macOS even for Docker Desktop setup. The downside
is that NFS does not support fsnotify events.

> To see how your project's Docker volumes are defined with `DOCKSAL_VOLUMES=nfs` see 

Reviewer: =nfs, see

[stacks/volumes-nfs.yml](https://github.com/docksal/docksal/blob/master/stacks/volumes-nfs.yml). 

### unison

> This option only works on macOS and Windows. Host files are made accessible to VM like in a bind mount, but after

Reviewer: to the VM

> that container does not access them directly via bind to avoid NFS/SMB performance penalty. Container file system is not 

Reviewer: that, the container... penalty. The container file system

connected to host filesystem at all.   

> Instead `fin` launches a `unison` container with Unison daemon that copies files back and forth between mounted host 

Reviewer: Instead,

files and the container. Transferrring changes from host to the container and back becomes slower, but reading and writing 

> the files within the container becomes way faster.

Reviewer: becomes much faster.

Author: what's wrong with "way faster"?

Reviewer: There is nothing “wrong” except that it’s more slang. “Much” would just be the preferred.

```
       NFS/SMB/osxfs                          Copy via Unison daemon
Host =================> VM/Docker Desktop < - - - - - - - - - - - - - > Container
```

The benefits of this setup:

- Full native container file system performance (reads and writes)
- `ionitify` event support **when Docker Desktop is used**

The downsides:

- Initial sync can take time, especially on large codebases
- Higher disk space usage (double the size of the codebase)
- Additional CPU load from the Unison daemon 

#### Using Unison Volumes

- Add `DOCKSAL_VOLUMES=unison` into `.docksal/docksal.env` of a project
- `fin project reset`
- Wait until initial sync finishes.

{{% notice note %}}

> Once you set a new `DOCKSAL_VOLUMES` option, you must re-create `cli` container. The easiest way is `fin project reset`,

Reviewer: recreate the cli container

> but it will also remove all data from the `db` volume. If you want to retain it then remove only `cli` container, and 

Reviewer: retain it, remove only the cli container

start the project again: `fin project remove cli; fin project start`
{{% /notice %}}

To see how your project's Docker volumes are defined with `DOCKSAL_VOLUMES=unison` see 
[stacks/volumes-unison.yml](https://github.com/docksal/docksal/blob/master/stacks/volumes-unison.yml).

### none

> Advanced option. With this option host files are made accessible to VM like in a bind mount, but after that 

Reviewer: With this option,
accessible to the VM
but after that,

> container does not access them directly via bind, nor files are copied over. 

Reviewer: the container

You would have to copy files into container and back manually.

> Combined with [VSCode IDE](/tools/ide/) this option can provide a way of provisioning instant blank 

Reviewer: ), this option

  - /en/master/advanced/volumes/
---

## Quick Overview

| OS      | VM              | DOCKSAL_VOLUMES              | Comments  |
|---------|-----------------|------------------------------|-----------|
| Linux   | -               | [bind](#bind) (**default**)  | Direct host files access, maximum filesystem speed. |

> | macOS   | VirtualBox      | [nfs](#nfs)  (**default**)   | **Pros:** fast, ~20% slower than native filesystem. <br> **Cons:** does not support filesystem events (fsnotify). |

Author: ```suggestion
| macOS   | VirtualBox      | [nfs](#nfs)  (**default**)   | **Pros:** fast, 80% of speed of the native filesystem. <br> **Cons:** does not support filesystem events (fsnotify). |
```

Мне всегда как-то странно читалась фраза "быстрый, на 20% медленнее". Предлагаю перефразировать.

Reviewer 2: Revamped the comparison table

| macOS   | Docker Desktop  | [nfs](#nfs)  (**default**)   | **Pros:** fast, ~20% slower than native filesystem. <br> **Cons:** does not support filesystem events (fsnotify). |
| macOS   | Docker Desktop  | [bind](#bind)                | **Pros:** supports filesystem events. <br> **Cons:** pretty slow, 40% slower than native filesystem. |

> | Windows | ANY             | [bind](#bind) (**default**)  | **Pros:** ~50% slower than native filesystem. <br> **Cons:** does not support filesystem events (fsnotify). |

Author: ```suggestion
| Windows | ANY             | [bind](#bind) (**default**)  | **Pros:** ~50% slower than native filesystem. <br> **Cons:** does not support filesystem events (fsnotify). |
```

Тут было 20%, откуда 50?

Reviewer 2: Revamped the comparison table

| macOS, Windows | ANY      | [unison](#unison)            | **Pros:** maximum `cli` filesystem performance. <br> **Cons:** initial wait for files to sync into `cli`; additional Docksal disk space use; sync delay when you switch git branches; higher CPU usage during files sync; sometimes Unison may 'break.' |
| ANY     | ANY             | [none](#none)                | **Pros:** maximum `cli` filesystem performance and no wait for the initial sync. <br> **Cons:** you have to copy files manually or checkout and edit files inside `cli` container. |

## Project Volumes

If you are familiar with Docker Compose, then you must have attached files or dirs like this:

```yaml
  cli:
    volumes:
      - /Users/alex/mysite:/var/www:rw
```

> That is a called a bind mount - mounting a files/folder on the host directly into a container. 

Author: ```suggestion
That is a called a bind mount - mounting a file/folder directly into a container. 
```

Let us avoid confusing user about what "host" is in terms of this documentation. You and I understand there are several "hosts", but to make the whole read clear to user in this doc lets call "host" a computer that runs Docker or VM. In this case it is not correct that bind mount example above would mount files from host, as on macOS, Windows it is not from host.

- `cli_home` stores home folder of your `cli` container independently of the `cli` container
- `db_data` stores your database data independently of the `db` container 
- `docksal_ssh_agent` stores SSH keys shared with all project stacks

Docksal makes these volumes to function transparently across different operating systems, but their actual definitions 
vary depending on the `DOCKSAL_VOLUMES` option.

{{% notice note %}}
`project_root` volume is mapped to a specific location on the host. 
If you need to move the location of a project on the computer, remove the project stack first with
`fin project rm`, then you can move the project folder and start the project stack again.
{{% /notice %}}

{{% notice warning %}}
When customizing `docksal.yml`, make sure to include the `cached` option **anywhere** the `project_root` volume is attached
to a service. Mixing `cached` and non-`cached` mounts for the same volume in your project stack will lead to issues 
and errors with Docker Desktop on macOS. See [docksal/docksal#678](https://github.com/docksal/docksal/issues/678) for more details.
{{% /notice %}}

## DOCKSAL_VOLUMES

> `DOCKSAL_VOLUMES` value changes the mount type for project volumes mentioned above and also affects `fin` behavior. 

Author: ```suggestion
`DOCKSAL_VOLUMES` value changes the mount type of the project volumes mentioned above and also affects `fin` behavior. 
```

> `DOCKSAL_VOLUMES` value changes the driver option for project volumes mentioned above and also affects `fin` behavior. 

Author: ```suggestion
`DOCKSAL_VOLUMES` value changes the mount type for project volumes mentioned above and also affects `fin` behavior. 
```

### bind 

With this option, containers access files via a [bind mount](https://docs.docker.com/storage/bind-mounts/), which basically means direct access. 

> While Docker can access files directly on Linux, on macOS and Windows it works inside a VM (VirtualBox, xhyve/Hyper-V 

Author: ```suggestion
While Docker can access files directly on Linux, on macOS and Windows it works inside the VM (VirtualBox, xhyve/Hyper-V 
```

As pre @shelane it is "the" here.

**On macOS with VirtualBox**, files are made available from host to Docker by mounting the folder defined 
in `DOCKSAL_NFS_PATH` into the VM via NFS protocol.
 
```
             NFS mount                     bind mount
macOS Host ==============> VirtualBox VM =============> Container       
```

**On macOS with Docker Desktop**, it is Docker Desktop itself that mounts folders defined in Docker Desktop UI via 
[osxfs](https://docs.docker.com/docker-for-mac/osxfs/) file system.

```
             osxfs mount                    bind mount
macOS Host ==============> Docker Desktop ==============> Container   
``` 

**On Windows with VirtualBox**, Docksal mounts **all** physical drives into the VM via SMB protocol.

```
               SMB mount                    bind mount
Windows Host =============> VirtualBox VM ==============> Container   
```

**On Windows with Docker Desktop**, it is Docker Desktop itself that mounts **only** configured Windows drives via SMB.

```
               SMB mount                     bind mount
Windows Host =============> Docker Desktop ==============> Container   
```

> To see how your project's Docker volumes are defined with when `DOCKSAL_VOLUMES=bind`, see 

Author: Remove with or when

## Quick Overview of the Volumes Options

If `DOCKSAL_VOLUMES` option is not marked as default, then you need to enable it manually to use it.

> All non-automatic options are enabled by placing `DOCKSAL_VOLUMES="<value from volumes column>"` into the respective

Reviewer: Should automatic be changed to default here too?

Author: Thanks, will change now

      - /Users/alex/mysite:/var/www:rw
```

> That is a called a bind mount - mounting a files/folder on the host directly into a container. 

Author: Let's return to what it was here. First the change didn't bring any new info. Second I deliberately avoided calling attaching volumes "mounting" to avoid confusing people who are new that everything is mounting. Mounting is used to describe a process of accessing host files from VM.

{{% notice warning %}}
When customizing `docksal.yml`, make sure to include the `cached` option **anywhere** the `project_root` volume is attached
to a service. Mixing `cached` and non-`cached` mounts for the same volume in your project stack will lead to issues 
and errors with Docker Desktop on macOS. See [docksal/docksal#678](https://github.com/docksal/docksal/issues/678) for more details.
{{% /notice %}}

## DOCKSAL_VOLUMES

`DOCKSAL_VOLUMES` value changes the driver option for project volumes mentioned above and also affects `fin` behavior. 

### bind 

> With this option, containers access files via a [bind mount](https://docs.docker.com/storage/bind-mounts/), which basically means direct access. 

Author: Again let's not call it mount.
